shim/src/main.rs

```diff
@@ -68,38 +84,120 @@ fn run_current_turbo(args: Vec<String>) -> Result<i32> {
+            anyhow!(
+                "Could not find binary in {}.",
+                local_turbo_path.to_string_lossy()
+            )
+        })?;
+
+        Ok(Some(local_turbo_path))
+    } else {
+        Ok(None)
     }
+}
 
-    // We should detect which package manager and then determine workspaces from there,
-    // but detection is not implemented yet and really we're either checking the `package.json`
-    // or the `pnpm-workspace.yaml` file so we can do both.
-    let npm = PackageManager::Npm;
-    if npm.get_workspace_globs(current_dir).is_ok() {
-        return Ok(false);
-    };
+impl RepoState {
+    /// Infers `RepoState` from current directory.
+    ///
+    /// # Arguments
+    ///
+    /// * `current_dir`: Current working directory
+    ///
+    /// returns: Result<RepoState, Error>
+    ///
+    pub fn infer(current_dir: &Path) -> Result<Self> {
+        // First we look for a `turbo.json`. This iterator returns the first ancestor that contains
+        // a `turbo.json` file.
+        let root_path = current_dir
+            .ancestors()
+            .find(|p| fs::metadata(p.join("turbo.json")).is_ok());
 
-    let pnpm = PackageManager::Pnpm;
-    if pnpm.get_workspace_globs(current_dir).is_ok() {
-        return Ok(false);
-    };
+        // If that directory exists, then we figure out if there are workspaces defined in it
+        // NOTE: This may change with multiple `turbo.json` files
+        if let Some(root_path) = root_path {
+            let pnpm = PackageManager::Pnpm;
+            let npm = PackageManager::Npm;
+            let is_workspace = pnpm.get_workspace_globs(root_path).is_ok()
+                || npm.get_workspace_globs(root_path).is_ok();
+
+            let mode = if is_workspace {
+                RepoMode::MultiPackage
+            } else {
+                RepoMode::SinglePackage
+            };
+
+            return Ok(Self {
+                root: root_path.to_path_buf(),
+                mode,
+            });
+        }
+
+        // What we look for next is a directory that contains a `package.json`.
+        let potential_roots = current_dir
+            .ancestors()
+            .filter(|path| fs::metadata(path.join("package.json")).is_ok());
+
+        let mut first_package_json_dir = None;
+        // We loop through these directories and see if there are workspaces defined in them,
+        // either in the `package.json` or `pnm-workspaces.yml`
+        for dir in potential_roots {
+            if first_package_json_dir.is_none() {
+                first_package_json_dir = Some(dir)
+            }
+
+            let pnpm = PackageManager::Pnpm;
+            let npm = PackageManager::Npm;
+            let is_workspace =
+                pnpm.get_workspace_globs(dir).is_ok() || npm.get_workspace_globs(dir).is_ok();
+
+            if is_workspace {
+                return Ok(Self {
+                    root: dir.to_path_buf(),
+                    mode: RepoMode::SinglePackage,
```

Comment:
Hmm yeah I was confused about this point. The doc says:

> ancestor path contains multiple package.json files → single package, rooted at closest package.json. This may be a monorepo, but with no turbo.json, we are forced to treat it as a single package.

But it does make sense that if there are workspaces, we can treat it as a multi package repo.